test/unit/transform/ferc1_test.py

```diff
@@ -642,79 +638,113 @@ def test_adding_parent_dimensions():
-            """
-table_name_parent,xbrl_factoid_parent,dim_x_parent,dim_y_parent,table_name,xbrl_factoid,dim_x,dim_y,is_within_dimension_dim_y,is_within_dimension_dim_x
-table_a,fact_1,voyager,coffee,table_a,fact_3,voyager,coffee,False,False
-table_a,fact_1,voyager,in,table_a,fact_3,voyager,in,False,False
-table_a,fact_1,voyager,that,table_a,fact_3,voyager,that,False,False
-table_a,fact_1,voyager,nebula,table_a,fact_3,voyager,nebula,False,False
-table_a,fact_1,voyager,total,table_a,fact_3,voyager,total,False,False
-table_a,fact_1,voyager,total,table_a,fact_1,voyager,coffee,True,False
-table_a,fact_1,voyager,total,table_a,fact_1,voyager,in,True,False
-table_a,fact_1,voyager,total,table_a,fact_1,voyager,that,True,False
-table_a,fact_1,voyager,total,table_a,fact_1,voyager,nebula,True,False
-table_a,fact_3,voyager,total,table_a,fact_3,voyager,coffee,True,False
-table_a,fact_3,voyager,total,table_a,fact_3,voyager,in,True,False
-table_a,fact_3,voyager,total,table_a,fact_3,voyager,that,True,False
-table_a,fact_3,voyager,total,table_a,fact_3,voyager,nebula,True,False
+    expected_total_to_subdim = canonicalize(
+        pd.read_csv(
+            StringIO(
+                """
+table_name_parent,xbrl_factoid_parent,dim_x_parent,dim_y_parent,table_name,xbrl_factoid,dim_x,dim_y,is_within_table_calc
+table_a,fact_1,voyager,coffee,table_a,fact_3,voyager,coffee,True
+table_a,fact_1,voyager,in,table_a,fact_3,voyager,in,True
+table_a,fact_1,voyager,that,table_a,fact_3,voyager,that,True
+table_a,fact_1,voyager,nebula,table_a,fact_3,voyager,nebula,True
+table_a,fact_1,voyager,total,table_a,fact_3,voyager,total,True
+table_a,fact_1,voyager,total,table_a,fact_1,voyager,coffee,NA
+table_a,fact_1,voyager,total,table_a,fact_1,voyager,in,NA
+table_a,fact_1,voyager,total,table_a,fact_1,voyager,that,NA
+table_a,fact_1,voyager,total,table_a,fact_1,voyager,nebula,NA
+table_a,fact_3,voyager,total,table_a,fact_3,voyager,coffee,NA
+table_a,fact_3,voyager,total,table_a,fact_3,voyager,in,NA
+table_a,fact_3,voyager,total,table_a,fact_3,voyager,that,NA
+table_a,fact_3,voyager,total,table_a,fact_3,voyager,nebula,NA
 """
+            )
+        )
+    )
+
+    out_total_to_subdim = canonicalize(
+        add_dimension_total_calculations(
+            calc_components=out_parent_dim_same_trek,
+            meta_w_dims=table_dimensions_same_trek,
+            table_dimensions=table_dimensions_same_trek,
+            dimensions=["dim_x", "dim_y"],
         )
     )
-    out_total_to_subdim = add_dimension_total_calculations(
-        calc_components=out_parent_dim_same_trek,
-        meta_w_dims=table_dimensions_same_trek,
-        table_dimensions=table_dimensions_same_trek,
-        dimensions=["dim_x", "dim_y"],
-    )[list(expected_total_to_subdim)]
 
     pd.testing.assert_frame_equal(
-        out_total_to_subdim.convert_dtypes(), expected_total_to_subdim.convert_dtypes()
+        out_total_to_subdim,
+        expected_total_to_subdim,
     )
 
 
 def test_multi_dims_totals():
```

Comment:
It would be sweet to generate these (relatively simple) data inputs with `hypothesis` but that seemed out of scope of this PR.

Reply:
I always find it difficult to understand what the structure of the test inputs are just by reading them. Does Hypothesis have strategies for generating tables with specific relational DB style constraints between the columns? e.g. many-to-one or one-to-one or ensuring some property holds true within groupbys? Or do you have some idea of how it would be easy to specify those kinds of relationships generically?

Reply:
What you would have to do is something like:

1. make a couple dataframes using the basic strategies for value generation (random choice from a predetermined list, random value within a range, etc.)
2. add a foreign key column to one dataframe which pulls values from the primary key column in the referent dataframe

We could maybe make some wrappers around that process to make it easier to use, or if we had some sort of SQLAlchemy <--> Pydantic <--> Pandas inter-op working, maybe we could use a plugin like [hypothesis-sqlalchemy](https://pypi.org/project/hypothesis-sqlalchemy/)

Reply:
In this scenario, I think we just need to draw stuff from the same possible value sets, e.g. `table_name` is always from "table_1", "table_2", etc. We could restrict table_dimensions to always be a subset of metadata, but :shrug: maybe it's good to know how our code would behave if something shows up in table_dimensions but not metadata.